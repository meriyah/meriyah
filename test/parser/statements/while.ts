import { outdent } from 'outdent';
import { describe } from 'vitest';
import { fail, pass } from '../../test-utils';

describe('Statements - While', () => {
  fail('Statements - While (fail)', [
    'while 1 break;',
    'while "hood" break;',
    'while (false) function f() {}',
    'while (false) let x = 1;',
    'while 1 break;',
    "while '' break;",
    { code: "while '' break;", options: { webcompat: true } },
    'while(0) !function(){ break; };',
    'while(0) { function f(){ break; } }',
    'while (false) label1: label2: function f() {}',
    'while (false) async function f() {}',
    'while (false) const x = null;',
    'while (false) function* g() {}',
    'while true break;',
    'while({1}){ break ; };',
    { code: 'while({1}){ break ; };', options: { webcompat: true } },
  ]);

  pass('Statements - While (pass)', [
    'while (1) /foo/',
    outdent`
      var i = 0;
      woohoo:{
        while(true){
          i++;
          if ( i == 10 ) {
            break woohoo;
          }
        }
      }
    `,
    outdent`
      while (false) let // ASI
      x = 1;
    `,
    outdent`
      while (false) let // ASI
      {}
    `,
    'while (x < 10) { x++; y--; }',
    'while (i-->1) {}',
    'a: while (true) continue a;',
    'while (this) try {} catch (h) {}',
    'while (foo) bar;',
  ]);
});
