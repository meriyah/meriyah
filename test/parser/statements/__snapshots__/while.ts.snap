// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Statements - While > Statements - While (fail) > while "hood" break; 1`] = `
"SyntaxError [1:6-1:12]: Expected '('
> 1 | while "hood" break;
    |       ^^^^^^ Expected '('"
`;

exports[`Statements - While > Statements - While (fail) > while '' break; 1`] = `
"SyntaxError [1:6-1:8]: Expected '('
> 1 | while '' break;
    |       ^^ Expected '('"
`;

exports[`Statements - While > Statements - While (fail) > while '' break; 2`] = `
"SyntaxError [1:6-1:8]: Expected '('
> 1 | while '' break;
    |       ^^ Expected '('"
`;

exports[`Statements - While > Statements - While (fail) > while (false) async function f() {} 1`] = `
"SyntaxError [1:20-1:28]: Async functions can only be declared at the top level or inside a block
> 1 | while (false) async function f() {}
    |                     ^^^^^^^^ Async functions can only be declared at the top level or inside a block"
`;

exports[`Statements - While > Statements - While (fail) > while (false) const x = null; 1`] = `
"SyntaxError [1:14-1:19]: Unexpected token: 'const'
> 1 | while (false) const x = null;
    |               ^^^^^ Unexpected token: 'const'"
`;

exports[`Statements - While > Statements - While (fail) > while (false) function f() {} 1`] = `
"SyntaxError [1:14-1:22]: Without web compatibility enabled functions can not be declared at top level, inside a block, or as the body of an if statement
> 1 | while (false) function f() {}
    |               ^^^^^^^^ Without web compatibility enabled functions can not be declared at top level, inside a block, or as the body of an if statement"
`;

exports[`Statements - While > Statements - While (fail) > while (false) function* g() {} 1`] = `
"SyntaxError [1:14-1:22]: Without web compatibility enabled functions can not be declared at top level, inside a block, or as the body of an if statement
> 1 | while (false) function* g() {}
    |               ^^^^^^^^ Without web compatibility enabled functions can not be declared at top level, inside a block, or as the body of an if statement"
`;

exports[`Statements - While > Statements - While (fail) > while (false) label1: label2: function f() {} 1`] = `
"SyntaxError [1:30-1:38]: Without web compatibility enabled functions can not be declared at top level, inside a block, or as the body of an if statement
> 1 | while (false) label1: label2: function f() {}
    |                               ^^^^^^^^ Without web compatibility enabled functions can not be declared at top level, inside a block, or as the body of an if statement"
`;

exports[`Statements - While > Statements - While (fail) > while (false) let x = 1; 1`] = `
"SyntaxError [1:18-1:19]: Unexpected token: 'identifier'
> 1 | while (false) let x = 1;
    |                   ^ Unexpected token: 'identifier'"
`;

exports[`Statements - While > Statements - While (fail) > while 1 break; 1`] = `
"SyntaxError [1:6-1:7]: Expected '('
> 1 | while 1 break;
    |       ^ Expected '('"
`;

exports[`Statements - While > Statements - While (fail) > while 1 break; 2`] = `
"SyntaxError [1:6-1:7]: Expected '('
> 1 | while 1 break;
    |       ^ Expected '('"
`;

exports[`Statements - While > Statements - While (fail) > while true break; 1`] = `
"SyntaxError [1:6-1:10]: Expected '('
> 1 | while true break;
    |       ^^^^ Expected '('"
`;

exports[`Statements - While > Statements - While (fail) > while({1}){ break ; }; 1`] = `
"SyntaxError [1:8-1:9]: Object literal keys that are strings or numbers must be a method or have a colon
> 1 | while({1}){ break ; };
    |         ^ Object literal keys that are strings or numbers must be a method or have a colon"
`;

exports[`Statements - While > Statements - While (fail) > while({1}){ break ; }; 2`] = `
"SyntaxError [1:8-1:9]: Object literal keys that are strings or numbers must be a method or have a colon
> 1 | while({1}){ break ; };
    |         ^ Object literal keys that are strings or numbers must be a method or have a colon"
`;

exports[`Statements - While > Statements - While (fail) > while(0) !function(){ break; }; 1`] = `
"SyntaxError [1:27-1:28]: Illegal break statement
> 1 | while(0) !function(){ break; };
    |                            ^ Illegal break statement"
`;

exports[`Statements - While > Statements - While (fail) > while(0) { function f(){ break; } } 1`] = `
"SyntaxError [1:30-1:31]: Illegal break statement
> 1 | while(0) { function f(){ break; } }
    |                               ^ Illegal break statement"
`;

exports[`Statements - While > Statements - While (pass) > a: while (true) continue a; 1`] = `
{
  "body": [
    {
      "body": {
        "body": {
          "label": {
            "name": "a",
            "type": "Identifier",
          },
          "type": "ContinueStatement",
        },
        "test": {
          "type": "Literal",
          "value": true,
        },
        "type": "WhileStatement",
      },
      "label": {
        "name": "a",
        "type": "Identifier",
      },
      "type": "LabeledStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Statements - While > Statements - While (pass) > var i = 0;
woohoo:{
  while(true){
    i++;
    if ( i == 10 ) {
      break woohoo;
    }
  }
} 1`] = `
{
  "body": [
    {
      "declarations": [
        {
          "id": {
            "name": "i",
            "type": "Identifier",
          },
          "init": {
            "type": "Literal",
            "value": 0,
          },
          "type": "VariableDeclarator",
        },
      ],
      "kind": "var",
      "type": "VariableDeclaration",
    },
    {
      "body": {
        "body": [
          {
            "body": {
              "body": [
                {
                  "expression": {
                    "argument": {
                      "name": "i",
                      "type": "Identifier",
                    },
                    "operator": "++",
                    "prefix": false,
                    "type": "UpdateExpression",
                  },
                  "type": "ExpressionStatement",
                },
                {
                  "alternate": null,
                  "consequent": {
                    "body": [
                      {
                        "label": {
                          "name": "woohoo",
                          "type": "Identifier",
                        },
                        "type": "BreakStatement",
                      },
                    ],
                    "type": "BlockStatement",
                  },
                  "test": {
                    "left": {
                      "name": "i",
                      "type": "Identifier",
                    },
                    "operator": "==",
                    "right": {
                      "type": "Literal",
                      "value": 10,
                    },
                    "type": "BinaryExpression",
                  },
                  "type": "IfStatement",
                },
              ],
              "type": "BlockStatement",
            },
            "test": {
              "type": "Literal",
              "value": true,
            },
            "type": "WhileStatement",
          },
        ],
        "type": "BlockStatement",
      },
      "label": {
        "name": "woohoo",
        "type": "Identifier",
      },
      "type": "LabeledStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Statements - While > Statements - While (pass) > while (1) /foo/ 1`] = `
{
  "body": [
    {
      "body": {
        "expression": {
          "regex": {
            "flags": "",
            "pattern": "foo",
          },
          "type": "Literal",
          "value": /foo/,
        },
        "type": "ExpressionStatement",
      },
      "test": {
        "type": "Literal",
        "value": 1,
      },
      "type": "WhileStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Statements - While > Statements - While (pass) > while (false) let // ASI
{} 1`] = `
{
  "body": [
    {
      "body": {
        "expression": {
          "name": "let",
          "type": "Identifier",
        },
        "type": "ExpressionStatement",
      },
      "test": {
        "type": "Literal",
        "value": false,
      },
      "type": "WhileStatement",
    },
    {
      "body": [],
      "type": "BlockStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Statements - While > Statements - While (pass) > while (false) let // ASI
x = 1; 1`] = `
{
  "body": [
    {
      "body": {
        "expression": {
          "name": "let",
          "type": "Identifier",
        },
        "type": "ExpressionStatement",
      },
      "test": {
        "type": "Literal",
        "value": false,
      },
      "type": "WhileStatement",
    },
    {
      "expression": {
        "left": {
          "name": "x",
          "type": "Identifier",
        },
        "operator": "=",
        "right": {
          "type": "Literal",
          "value": 1,
        },
        "type": "AssignmentExpression",
      },
      "type": "ExpressionStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Statements - While > Statements - While (pass) > while (foo) bar; 1`] = `
{
  "body": [
    {
      "body": {
        "expression": {
          "name": "bar",
          "type": "Identifier",
        },
        "type": "ExpressionStatement",
      },
      "test": {
        "name": "foo",
        "type": "Identifier",
      },
      "type": "WhileStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Statements - While > Statements - While (pass) > while (i-->1) {} 1`] = `
{
  "body": [
    {
      "body": {
        "body": [],
        "type": "BlockStatement",
      },
      "test": {
        "left": {
          "argument": {
            "name": "i",
            "type": "Identifier",
          },
          "operator": "--",
          "prefix": false,
          "type": "UpdateExpression",
        },
        "operator": ">",
        "right": {
          "type": "Literal",
          "value": 1,
        },
        "type": "BinaryExpression",
      },
      "type": "WhileStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Statements - While > Statements - While (pass) > while (this) try {} catch (h) {} 1`] = `
{
  "body": [
    {
      "body": {
        "block": {
          "body": [],
          "type": "BlockStatement",
        },
        "finalizer": null,
        "handler": {
          "body": {
            "body": [],
            "type": "BlockStatement",
          },
          "param": {
            "name": "h",
            "type": "Identifier",
          },
          "type": "CatchClause",
        },
        "type": "TryStatement",
      },
      "test": {
        "type": "ThisExpression",
      },
      "type": "WhileStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;

exports[`Statements - While > Statements - While (pass) > while (x < 10) { x++; y--; } 1`] = `
{
  "body": [
    {
      "body": {
        "body": [
          {
            "expression": {
              "argument": {
                "name": "x",
                "type": "Identifier",
              },
              "operator": "++",
              "prefix": false,
              "type": "UpdateExpression",
            },
            "type": "ExpressionStatement",
          },
          {
            "expression": {
              "argument": {
                "name": "y",
                "type": "Identifier",
              },
              "operator": "--",
              "prefix": false,
              "type": "UpdateExpression",
            },
            "type": "ExpressionStatement",
          },
        ],
        "type": "BlockStatement",
      },
      "test": {
        "left": {
          "name": "x",
          "type": "Identifier",
        },
        "operator": "<",
        "right": {
          "type": "Literal",
          "value": 10,
        },
        "type": "BinaryExpression",
      },
      "type": "WhileStatement",
    },
  ],
  "sourceType": "script",
  "type": "Program",
}
`;
